# 1.A

In [2]:
import pandas as pd
from glob import glob

In [3]:
p_file=sorted(glob('//Users//debanarzary//Downloads//Data Set//phone_user_review_file_*.csv'))
p_file

['//Users//debanarzary//Downloads//Data Set/phone_user_review_file_1.csv',
 '//Users//debanarzary//Downloads//Data Set/phone_user_review_file_2.csv',
 '//Users//debanarzary//Downloads//Data Set/phone_user_review_file_3.csv',
 '//Users//debanarzary//Downloads//Data Set/phone_user_review_file_4.csv',
 '//Users//debanarzary//Downloads//Data Set/phone_user_review_file_5.csv',
 '//Users//debanarzary//Downloads//Data Set/phone_user_review_file_6.csv']

In [4]:
df=pd.concat((pd.read_csv(file,encoding='latin-1').assign(filename=file)
             for file in p_file),ignore_index=True)

# 1.B

In [5]:
df.describe()

score  score_max
count  1.351644e+06  1351644.0
mean   8.007060e+00       10.0
std    2.616121e+00        0.0
min    2.000000e-01       10.0
25%    7.200000e+00       10.0
50%    9.200000e+00       10.0
75%    1.000000e+01       10.0
max    1.000000e+01       10.0

In [6]:
df.dtypes

phone_url     object
date          object
lang          object
country       object
source        object
domain        object
score        float64
score_max    float64
extract       object
author        object
product       object
filename      object
dtype: object

In [7]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
filename         0
dtype: int64

Observations:

1)There are only 2 floats and rest are objects in the dataframe.

2)There are lots of missing values in the dataframe.

# 1.C

In [8]:
df['score']=round(df['score'])
df['score']

0          10.0
1          10.0
2           6.0
3           9.0
4           4.0
           ... 
1415128     2.0
1415129    10.0
1415130     2.0
1415131     8.0
1415132     2.0
Name: score, Length: 1415133, dtype: float64

In [9]:
df.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.0       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  \
0                                  Samsung Galaxy S8   
1                                  Samsung Galaxy S8   
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...   
3                      Samsung Galaxy S8 64GB (AT&T)   
4                                  Samsung Galaxy S8   

                                            filename  
0  //Users//debanarzary//Downloads//Data Set/phon...  
1  //Users//debanarzary//Downloads//Data Set/phon...  
2  //Users//debanarzary//Downloads//Data Set/phon...  
3  //Users//debanarzary//Downloads//Data Set/phon...  
4  //Users//debanarzary//Downloads//Data Set/phon...

# 1.D

In [10]:
df.isna().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
filename         0
dtype: int64

In [11]:
df['score']=df['score'].fillna(df['score'].mean())
df['score']

0          10.0
1          10.0
2           6.0
3           9.0
4           4.0
           ... 
1415128     2.0
1415129    10.0
1415130     2.0
1415131     8.0
1415132     2.0
Name: score, Length: 1415133, dtype: float64

In [12]:
df['score_max']=df['score_max'].fillna(df['score_max'].mean())
df['score_max']

0          10.0
1          10.0
2          10.0
3          10.0
4          10.0
           ... 
1415128    10.0
1415129    10.0
1415130    10.0
1415131    10.0
1415132    10.0
Name: score_max, Length: 1415133, dtype: float64

In [13]:
df=df.dropna(axis=0)

In [14]:
df.shape

(1336416, 12)

# 1.E

In [15]:
df.duplicated().any()

True

In [16]:
df=df.drop_duplicates()

In [17]:
df.shape

(1331847, 12)

# 1.F

In [18]:
df.sample(n=1000000,axis=0,random_state=612)

phone_url        date lang country  \
551832             /cellphones/huawei-ascend-g6/   9/16/2014   ru      ru   
554856           /cellphones/blackberry-classic/    7/6/2016   de      de   
868886                /cellphones/sony-xperia-e/   19/1/2015   en      gb   
547724                  /cellphones/oneplus-two/   9/13/2015   en      in   
589444  /cellphones/samsung-g355-galaxy-core-ii/   1/14/2016   it      it   
...                                          ...         ...  ...     ...   
197280            /cellphones/samsung-galaxy-a5/   4/13/2015   en      us   
433208         /cellphones/panasonic-eluga-mark/  12/21/2015   en      in   
786065               /cellphones/wiko-cink-slim/   3/26/2013   fr      fr   
230971       /cellphones/microsoft-lumia-950-xl/   1/16/2016   it      it   
384235               /cellphones/sony-xperia-z3/   10/3/2014   es      es   

               source              domain  score  score_max  \
551832        Onliner  catalog.onliner.by   10.0       10.0   
554856         Amazon           amazon.de    8.0       10.0   
868886         Amazon        amazon.co.uk   10.0       10.0   
547724         Amazon           amazon.in    2.0       10.0   
589444         Amazon           amazon.it    8.0       10.0   
...               ...                 ...    ...        ...   
197280         Amazon          amazon.com    8.0       10.0   
433208         Amazon           amazon.in   10.0       10.0   
786065  Mobiles-Actus   mobiles-actus.com    6.0       10.0   
230971         Amazon           amazon.it    2.0       10.0   
384235         Amazon           amazon.es    8.0       10.0   

                                                  extract           author  \
551832  2 недели пользуюсь данным...            losza   
554856  In den letzten zweieinhalb Jahren war ich (bis...        piquadrat   
868886       Nice phone good value good service thank you     cheryl trigg   
547724  I am using this device for 4-5 months now. I w...  Amazon Customer   
589444  Il terminale ha soddisfatto le aspettative di ...         Marcello   
...                                                   ...              ...   
197280  This phone is another Samsung slam dunk. Howev...           Meesha   
433208                                Best config + Style         Shirshak   
786065  TROP tôt pour dire un peu compliquer, je comp...         gislaine   
230971  Recapitato con precisione chirurgica. Telefono...        Medico118   
384235  Teléfono excepcional con una calidad y diseñ...           Emilio   

                                                  product  \
551832                               Huawei Ascend G6-U10   
554856  Blackberry Classic Smartphone (3,5 Zoll (8,9 c...   
868886  Xperia Sony Xperia E SIM-free Android Smartpho...   
547724                  OnePlus 2 (Sandstone Black, 64GB)   
589444  Samsung G355HN Galaxy Core 2 Duos Smartphone, ...   
...                                                   ...   
197280  Samsung Galaxy A5 LTE white SM-A500F Factory U...   
433208      Panasonic Eluga Mark EB-90S55EMK (Royal Gold)   
786065                                     Wiko Cink Slim   
230971  Microsoft Telefonia Lumia 950 XL Smartphone, 3...   
384235  Sony Xperia Z3 - Smartphone (13,21 cm (5.2"), ...   

                                                 filename  
551832  //Users//debanarzary//Downloads//Data Set/phon...  
554856  //Users//debanarzary//Downloads//Data Set/phon...  
868886  //Users//debanarzary//Downloads//Data Set/phon...  
547724  //Users//debanarzary//Downloads//Data Set/phon...  
589444  //Users//debanarzary//Downloads//Data Set/phon...  
...                                                   ...  
197280  //Users//debanarzary//Downloads//Data Set/phon...  
433208  //Users//debanarzary//Downloads//Data Set/phon...  
786065  //Users//debanarzary//Downloads//Data Set/phon...  
230971  //Users//debanarzary//Downloads//Data Set/phon...  
384235  //Users//debanarzar

# 1.G

In [19]:
df=df.drop(['phone_url','date','lang','country','source','domain','score_max','extract','filename'],axis=1)

In [20]:
df.head()

score       author                                            product
0   10.0   CarolAnn35                                  Samsung Galaxy S8
1   10.0    james0923                                  Samsung Galaxy S8
2    6.0     R. Craig  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...
3    9.0  Buster2020                       Samsung Galaxy S8 64GB (AT&T)
4    4.0   S Ate Mine                                  Samsung Galaxy S8

# 2.A

In [21]:
product_rating=df.groupby('product')['score'].count().sort_values(ascending=False).reset_index()
product_rating.rename(columns={'score':'num_rating'},inplace=True)
product_rating

product  num_rating
0                       Lenovo Vibe K4 Note (White,16GB)        5223
1                      Lenovo Vibe K4 Note (Black, 16GB)        4389
2                            OnePlus 3 (Graphite, 64 GB)        4103
3                           OnePlus 3 (Soft Gold, 64 GB)        3557
4                            Huawei P8lite zwart / 16 GB        2707
...                                                  ...         ...
59989                            Nokia ExpressMusic 5130           1
59990                           Nokia Express Music 5130           1
59991            Nokia EE Nokia 130 Mobile Phone - Black           1
59992  Nokia E90 Communicator Smartphone HSDPA/quadri...           1
59993  金屬 + 皮革 = 質感滿分　Wave 723 詳æ...           1

[59994 rows x 2 columns]

# 2.B

In [22]:
user_rating=df.groupby('author')['score'].count().sort_values(ascending=False).reset_index()
user_rating.rename(columns={'score':'user_rating'},inplace=True)
user_rating

author  user_rating
0                                         Amazon Customer        76950
1                                          Cliente Amazon        19300
2                                                   e-bit         8411
3                                         Client d'Amazon         7716
4                                            Amazon Kunde         4746
...                                                   ...          ...
792607                                      Lyda Trujillo            1
792608                                             Lyde10            1
792609                                          Lydi Rosi            1
792610                                          Lydia Ann            1
792611  ��������� X�� ���ï...            1

[792612 rows x 2 columns]

# 2.C

In [23]:
product=product_rating[product_rating['num_rating']>50]
product

product  num_rating
0                      Lenovo Vibe K4 Note (White,16GB)        5223
1                     Lenovo Vibe K4 Note (Black, 16GB)        4389
2                           OnePlus 3 (Graphite, 64 GB)        4103
3                          OnePlus 3 (Soft Gold, 64 GB)        3557
4                           Huawei P8lite zwart / 16 GB        2707
...                                                 ...         ...
6253     Elephone Batería Original para Elephone P7000          51
6254     Motorola DROID RAZR MAXX HD (Verizon Wireless)          51
6255  Motorola DROID RAZR MAXX, Black 32GB (Verizon ...          51
6256  Samsung Galaxy S7 32 GB Unlocked Phone - G930F...          51
6257                                         Nokia 8910          51

[6258 rows x 2 columns]

In [24]:
user=user_rating[user_rating['user_rating']>50]
user

author  user_rating
0    Amazon Customer        76950
1     Cliente Amazon        19300
2              e-bit         8411
3    Client d'Amazon         7716
4       Amazon Kunde         4746
..               ...          ...
920       Диана           51
921          dik1609           51
922         Rezultat           51
923          richard           51
924          Vicente           51

[925 rows x 2 columns]

In [25]:
product.shape

(6258, 2)

In [26]:
user.shape

(925, 2)

# 3

In [27]:
avg_rating=df.groupby('product')['score'].mean().sort_values(ascending=False).reset_index()
avg_rating.rename(columns={'score':'avg_rating'},inplace=True)
avg_rating

product  avg_rating
0      Nokia 5310 XpressMusic Rose & Noir - Télépho...        10.0
1                                 Sony Xperia Z1 - Black        10.0
2      LG G4 H815 32GB 4G-Smartphone Movistar Libre,(...        10.0
3       BlackBerry Curve 3G 9300 Red Unlocked Smartphone        10.0
4      LG G4 H815 32GB 4G-Smartphone Libre(13,97 cm (...        10.0
...                                                  ...         ...
59989                             Sony Xperia Z2 (black)         1.0
59990  Lenovo Motorola Devour- silver (Verizon Wireless)         1.0
59991         Kyocera Loft S2300 - black (Virgin Mobile)         1.0
59992                        BlackBerry Bold 9700 (AT&T)         1.0
59993                     Samsung Mythic SGH-A897 (AT&T)         1.0

[59994 rows x 2 columns]

In [28]:
popular_df=product_rating.merge(avg_rating,on='product')
popular_df=popular_df[popular_df['num_rating']>50].sort_values('avg_rating',ascending=False)
popular_df.head(5)

product  num_rating  \
1218  Samsung Smartphone Dual Chip Samsung Galaxy SI...         153   
1024  Samsung Smartphone Dual Chip Samsung Galaxy SI...         171   
1283  Motorola Smartphone Motorola Novo Moto G DTV C...         147   
1155  Samsung Smartphone Galaxy Win Duos Branco Desb...         158   
1032  Nokia Smartphone Nokia Lumia 520 Desbloqueado ...         170   

      avg_rating  
1218        10.0  
1024        10.0  
1283        10.0  
1155        10.0  
1032        10.0

# 4

In [29]:
dff=df.sample(n=5000,axis=0)

In [30]:
from surprise import Dataset,Reader

reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(dff[['author', 'product', 'score']], reader)

In [31]:
from surprise.model_selection import train_test_split
trainset, testset = train_test_split(data, test_size=.25,random_state=123)

In [32]:
trainset.all_ratings()

<generator object Trainset.all_ratings at 0x7f7ba809a0b0>

In [33]:
from surprise import SVD, KNNWithMeans
from surprise import accuracy

In [34]:
svd_model = SVD(n_factors=5,biased=False)
svd_model.fit(trainset)

In [35]:
testset[0]

('Aakash Naskar', 'LG Optimus Hub', 2.0)

In [36]:
test_pred = svd_model.test(testset)

In [37]:
accuracy.rmse(test_pred)

RMSE: 2.9589


2.95887502422136

In [38]:
from surprise import KNNWithMeans
from surprise import accuracy
algo_i = KNNWithMeans(k=10, sim_options={ 'user_based': False})
algo_i.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [39]:
test_pred=algo_i.test(testset)
print(accuracy.rmse(test_pred))

RMSE: 2.6415
2.64145855129507


# 5

In [40]:
accuracy.rmse(test_pred)

RMSE: 2.6415


2.64145855129507

# 6

In [49]:
test_user=pd.DataFrame(test_pred)
test_user=test_user[['uid','iid','r_ui','est']]
test_user.rename(columns={'uid':'t_user','iid':'product','r_ui':'original_rating','est':'estimation'},inplace=True)
t_user=test_user.drop_duplicates()
t_user

t_user                                            product  \
0          Aakash Naskar                                     LG Optimus Hub   
1     александр       Samsung Galaxy S5 SM-G900F 16GB (синий)   
2           LETERTRE Guy                                       Samsung J750   
3                    RW1  Sim Free Apple iPhone SE 64GB Mobile Phone - R...   
4                   Juan  Nokia E63-2 Unlocked Phone with 2 MP Camera, 3...   
...                  ...                                                ...   
1245          T. mcneill             LG Flare Prepaid Phone (Virgin Mobile)   
1246     Prashanth Nayak                  Apple iPhone 6 (Space Grey, 64GB)   
1247             stefano  LG X Power Smartphone, Schermo IPS 5.3" HD, 4G...   
1248               Danaa  Sony Smartphone Sony Xperia M Dual Desbloquead...   
1249      Cliente Amazon  Doogee X5 - Smartphone Libre 3G Android 5.1 (Q...   

      original_rating  estimation  
0            2.000000    7.994728  
1            2.000000    7.994728  
2            8.008083    7.994728  
3           10.000000    7.994728  
4           10.000000    7.994728  
...               ...         ...  
1245        10.000000    7.994728  
1246        10.000000    7.994728  
1247         8.000000    7.994728  
1248         6.000000    7.994728  
1249        10.000000   10.000000  

[1249 rows x 4 columns]

In [53]:
t_user_avg_rating=pd.DataFrame(t_user.groupby('product')['estimation'].mean().sort_values(ascending=False))
t_user_avg_rating

estimation
product                                                       
Doogee X5 - Smartphone Libre 3G Android 5.1 (Qu...   10.000000
OnePlus X (Limited Edition Ceramic)                  10.000000
DOOGEE T6 Pro Smartphone 4G - 5.5 IPS Android 6...   10.000000
Samsung T404G Prepaid Phone With Triple Minutes...   10.000000
Apple iPhone 4 CDMA Verizon Cellphone, 16GB, Black   10.000000
...                                                        ...
Smartphone Samsung Galaxy Young 2 SM-G130             2.000000
Apple iPhone 6 Plus GSM Unlocked Cellphone, 64G...    2.000000
HTC Desire 526G Plus (Fervor Red, 16GB)               2.000000
Lenovo K3 Note (Black)                                1.700000
Microsoft Nokia Lumia 630 Single-SIM Smartphone...    1.203959

[1123 rows x 1 columns]

# 7

Findings & Inferences:

1)There are 1123 test users estimated ratings across all products.

2)Estimation rating can be less than or more than the original rating.

# 8

In [54]:
t_user.sort_values('estimation',ascending=False).head(5) #Top 5 products for test users

t_user                                            product  \
1249  Cliente Amazon  Doogee X5 - Smartphone Libre 3G Android 5.1 (Q...   
1234    einem Kunden  Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,...   
84          Иван              Sony Xperia Z1 (фиолетовый)   
1162       Stephanie  Samsung Galaxy S6 64 GB UK SIM-Free Smartphone...   
1156          Walter  Honor 6X Smartphone portable débloqué 4G (Ec...   

      original_rating  estimation  
1249             10.0        10.0  
1234              8.0        10.0  
84                9.0        10.0  
1162             10.0        10.0  
1156             10.0        10.0

# 9

In [55]:
from surprise.model_selection import cross_validate
cross_validate(svd_model, data, measures=["RMSE", "MAE"], cv=5, verbose=True) #SVD cross validation

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.9764  2.8466  2.9458  2.7355  2.9040  2.8817  0.0851  
MAE (testset)     2.1942  2.0949  2.1724  2.0720  2.1498  2.1367  0.0462  
Fit time          0.14    0.07    0.06    0.07    0.08    0.08    0.03    
Test time         0.04    0.02    0.02    0.02    0.02    0.02    0.01    


{'test_rmse': array([2.9763839 , 2.84661816, 2.94577554, 2.7355176 , 2.90401202]),
 'test_mae': array([2.19424869, 2.09493007, 2.17244994, 2.07198922, 2.14979457]),
 'fit_time': (0.13540983200073242,
  0.06994199752807617,
  0.05936884880065918,
  0.068939208984375,
  0.08430266380310059),
 'test_time': (0.0400700569152832,
  0.018551111221313477,
  0.018638134002685547,
  0.018732070922851562,
  0.015646934509277344)}

In [56]:
cross_validate(algo_i, data, measures=["RMSE", "MAE"], cv=5, verbose=True) #KNNWithMeans cross validation

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.5922  2.5838  2.6460  2.5481  2.7030  2.6146  0.0542  
MAE (testset)     1.9854  1.9362  1.9953  1.9477  2.0273  1.9784  0.0330  
Fit time          0.43    0.30    0.29    0.29    0.40    0.34    0.06    
Test time         0.03    0.04    0.03    0.03    0.03    0.03    0.00    


{'test_rmse': array([2.59223472, 2.58380414, 2.64595431, 2.54814363, 2.70300194]),
 'test_mae': array([1.98543637, 1.93623532, 1.99530083, 1.94773409, 2.02731913]),
 'fit_time': (0.43492817878723145,
  0.30059814453125,
  0.28601884841918945,
  0.29436802864074707,
  0.4012937545776367),
 'test_time': (0.02917933464050293,
  0.039260149002075195,
  0.02519989013671875,
  0.03395819664001465,
  0.032302141189575195)}

# 10

Business scenario for using popularity based Recommended System:

When Disney plus hotstar app has to recommend movies which are popular and trending.

# 11

Business scenario for using CF based Recommendation Systems:
 
When Netflix app has to recommend new TV shows for a specific genre oriented customer.

# 12

For further improvement of the recommendation for different users, hybrid recommender can be used.